### ISSUE TITLE
Edit menu Paste is NOT the same as a Ctrl+v paste

### ISSUE THREAD

**Issue by sasumner** [notepad-plus-plus/notepad-plus-plus#9260]
### Description of the Issue

The _Edit_ menu has an entry called _Paste_ and by default keymapping is shown as mapping to Ctrl+v and Shift+Ins:

![image](https://user-images.githubusercontent.com/30118311/102012536-0efefa00-3d19-11eb-970e-b27ccba821ce.png)

However, the actions performed when the menu item is invoked from the menu are different from the actions performed when Ctrl+v is pressed.

This seems wrong because from what appears on the _Edit_ menu, the commands are one-and-the-same.

### Steps to Reproduce the Issue

1. Turn visible line endings on by selecting _View_ menu > _Show Symbol_ > _Show End of Line_
2. Create some data in a tab; make the tab be of "Windows (CR LF)" line-ending type:
![image](https://user-images.githubusercontent.com/30118311/102012730-212d6800-3d1a-11eb-8cb9-f4c169415fe5.png)
3. Change some or all line-endings in this tab to "LF" by a regular expression replacement of `\R` with `\n` (okay, not something normally done, but this is for point of illustration that the menu item paste is doing something different than Ctrl+v); observe:
![image](https://user-images.githubusercontent.com/30118311/102012879-b4ff3400-3d1a-11eb-89d4-32f287318e3a.png)
4. Select lines 3 and 4 and press Ctrl+c ; move to end of document and press Ctrl+v ; observe:
![image](https://user-images.githubusercontent.com/30118311/102012921-f8f23900-3d1a-11eb-86c3-1df91386220b.png)
5. Press Ctrl+z to undo the actions of step 4 and return to the state shown in step 3.
6. Select lines 3 and 4 and press Ctrl+c ; move to end of document and invoke _Paste_ from the _Edit_ menu ; observe:
![image](https://user-images.githubusercontent.com/30118311/102012972-335bd600-3d1b-11eb-8b93-80f130a34dea.png)

### Expected Behavior

A consistent paste action whether the _Paste_ menu entry is invoked, or Ctrl+v is pressed.
In other words, the line-endings after step 4 and after step 6 should be the same.

### Actual Behavior

Different line-endings result after step 4 and step 6.
Specifically, the _Edit_ menu _Paste_ has changed all lines of the file in addition to pasting the desired data, whereas Ctrl+v has only pasted the desired data.

### Other

Note that a right-click context menu _Paste_ produces the same result as choosing the _Edit_ menu _Paste_ command. 

[HERE](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/a0ffb6b6ff8cef58ad7d03a8893ef69560fed63e/PowerEditor/src/NppCommands.cpp#L380) is the code for `IDM_EDIT_PASTE` which is the _Edit_ menu invocation version.  Presume that the Ctrl+v version is simply doing an `SCI_PASTE`.

### Debug Information
Notepad++ v7.9.1   (64-bit)
Build time : Nov  2 2020 - 01:07:46
Path : C:\...............\npp.7.9.1.portable.x64\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 1809
OS Build : 17763.1577
Current ANSI codepage : 1252
Plugins : mimeTools.dll NppConverter.dll NppExport.dll

**Comment by ArkadiuszMichalski** [comment]
What other editors doing here? I see that SciTE also convert to actual line-endings. VSC and Sublime Text don't show this chars in-build, so I can't verify.

**Comment by donho** [comment]
Out of topic: 
Since when undo, redo, copy, cut and paste have 2 shortcuts?

**Comment by sasumner** [comment]
>Since when undo, redo, copy, cut and paste have 2 shortcuts?

Isn't the author of the program the best person to answer this? :-)
AFAIK, Scintilla commands have always (my "always" being about 5 years!) been able to have 2 different shortcuts.

**Comment by donho** [comment]
> Isn't the author of the program the best person to answer this? :-)

Damn! The modification is at least 5 years ago. I don't really remember this part of change...
With "git blame" I can only get f**king useless commit (https://github.com/notepad-plus-plus/notepad-plus-plus/blame/master/PowerEditor/src/Parameters.cpp#L432) for reformatting by a contributor.

**Comment by donho** [comment]
> Different line-endings result after step 4 and step 6.

Hein? From your screenshots, both lines in step 4 & step 6 are `CR LF`.

**Comment by sasumner** [comment]
>both lines in step 4 & step 6 are CR LF.

???  I don't see it.  I'm comparing these two text file contents:

![image](https://user-images.githubusercontent.com/30118311/102302349-dfbdd800-3f26-11eb-8d27-01d1c8d2874b.png)

and

![image](https://user-images.githubusercontent.com/30118311/102302365-ecdac700-3f26-11eb-8be3-ba8a97de1414.png)

I put great effort into typing up issues so that they are very clear.  Maybe I fail.  :-(

**Comment by donho** [comment]
No, sorry but I read again the instructions, still vague to me :(

**Comment by Ekopalypse** [comment]
@Donho - the resulting lines 7 and 8 are the same, but the original lines 1-6 are different.
CTRL+V does NOT change the line endings of 1-6, whereas pasting via the menu changes the line endings for lines 1-6.

**Comment by donho** [comment]
@Ekopalypse Thank you for your explanation.
@sasumner I just cannot reproduce it.
Copy 2 lines from a `CR LF`, and running Ctrl-V and "Edit->Paste" have the same result on `CR LF` and `LF` : on the `CR LF` document the pasted lines with `CR LF` and on the `LF` document the pasted lines with `LF`.
Copy 2 lines from a `LF`, and running Ctrl-V and "Edit->Paste" have the same result.

**Comment by Ekopalypse** [comment]
I assume here is the misunderstanding  _"... on the LF document..."_
The document is ALWAYS a CRLF document.

**Comment by donho** [comment]
@Ekopalypse  So you can reproduce what @sasumner has reproduced?
What's your steps to reproduce this issue?

**Comment by Ekopalypse** [comment]
I did the same that @sasumner did.
But I find it easier doing the following to show the same result.
Insert the following line into npp with the line ends having CRLF
~~~
test1test2test3test4test5CRLF

~~~
Now place the cursor after number 1 and press enter
Do same after the number 4
Result:
~~~
test1CRLF
test2test3test4CRLF
test5CRLF
~~~
Open the character panel
place the cursor after number 2 and insert a LF from character panel
Do the same after number 3
Result;
~~~
test1CRLF
test2LF
test3LF
test4CRLF
test5CRLF
~~~

Now comes the thing why the issue was opened

select the line test2 and test3 and copy it
goto the end of the document und paste it using ctrl + v
result:
~~~
test1CRLF
test2LF
test3LF
test4CRLF
test5CRLF
test2CRLF
test3CRLF
~~~

UNDO this copy and use paste from either the main menu or the context menu
Result:
~~~
test1CRLF
test2CRLF
test3CRLF
test4CRLF
test5CRLF
test2CRLF
test3CRLF
~~~

**Comment by Ekopalypse** [comment]
Like this
![paste](https://user-images.githubusercontent.com/47723516/102414487-ed429300-3ff6-11eb-911e-b210d227581e.gif)

**Comment by donho** [comment]
@Ekopalypse From your animated gif, "Ctrl-V" & "Edit->Paste" insert both "<text>CR LF" even you copied "<text>LF", right?
In this case, I'm confused by the issue's title: `Edit menu Paste is NOT the same as a Ctrl+v paste`.

But it's an "expected" behaviour for me, even there's a glitch in this case:
If the paste operation is applied to a document judged as Windows document (all EOL are "CR LF"), the EOL of pasted content will be modified to "CR LF".

**Comment by donho** [comment]
@Ekopalypse OK, I see.

"Paste->Edit" is supposed to take over "CTRL-V", but it doesn't.
MY question here is: do you guys think it's a good behaviour for "Paste->Edit" ?

**Comment by sasumner** [comment]
@donho 

>MY question here is: do you guys think it's a good behaviour for "Paste->Edit" ?

I believe that we are still thinking on it. :-)

My goal in creating the issue was to point out the inconsistency in the 2 different actions, which is unexpected.

BUT:
I think it might be unexpected for a paste action to affect the document globally (by converting ALL of the document's EOLs).
So that is a "bigger" issue.

**Comment by Ekopalypse** [comment]
I agree, even if it is only the end-of-line characters, I, personally, would be surprised to see them change due to a copy/cut/paste operation elsewhere in the document. 
From my point of view, what is obtained via a copy/cut should also be reinserted by the paste action, even if this means that it does not correspond to what is defined as the end-of-line character for the document, BUT this does not seem to be supported by Scintilla. Honestly, I'm not sure if there is a use case for mixed EOLs at all. 
The goal should be to have one behaviour for both actions, whether this is called global change or local change only is secondary in my view. There are probably as many reasons to do one as the other.

**Comment by sasumner** [comment]
>what is obtained via a copy/cut should also be reinserted by the paste action, even if this means that it does not correspond to what is defined as the end-of-line character for the document, BUT this does not seem to be supported by Scintilla.

It is [supported by Scintilla](https://www.scintilla.org/ScintillaDoc.html#SCI_SETPASTECONVERTENDINGS), but Notepad++ has made the choice (maybe it was a choice not to choose!) not to do it that way; proof:

![image](https://user-images.githubusercontent.com/30118311/102508812-d5e5c300-4053-11eb-8e06-e00fd77bf7f5.png)

By "a choice not to choose", I mean that currently no call to `SCI_SETPASTECONVERTENDINGS` exists in the N++ source code, and from the Scintilla docs this feature "Defaults to true."

>I'm not sure if there is a use case for mixed EOLs at all.

I totally agree.  But probably others would disagree (e.g. people interested in #7735).

**Comment by Ekopalypse** [comment]
Nice find, so a config setting like "disallow paste converting eols" and making pasting via the menu consistent with CTRL+V would make almost everyone happy, right !?

**Comment by donho** [comment]
> Nice find, so a config setting like "disallow paste converting eols" and making pasting via the menu consistent with CTRL+V would make almost everyone happy, right !?

Once again, "too many options, no option".

**Comment by ArkadiuszMichalski** [comment]
At least we should correct naming (or change behaviour) for `Paste` in the `Edit` menu because now it's confusing.

`Ctrl+V` is bind to `SCI_PASTE` and actually Notepad++ convert EOL (**only in pasted data**) to what we have set for document. This conversion is done because it's Scintilla default behavior. We can control this by [`SCI_SETPASTECONVERTENDINGS`](https://www.scintilla.org/ScintillaDoc.html#SCI_SETPASTECONVERTENDINGS).
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/bdb06d5c109f63eac5ab908a7c4f2718fdcc634d/PowerEditor/src/Parameters.cpp#L458-L459
![image](https://user-images.githubusercontent.com/2730894/211419675-4f7315c9-a1b1-43a8-804d-e9496b5ea4ff.png)

`Edit > Paste` is bind to `IDM_EDIT_PASTE` and this command also use  `SCI_PASTE`, but additionally convert EOL  (**in all document**) by using [SCI_CONVERTEOLS]( https://www.scintilla.org/ScintillaDoc.html#SCI_CONVERTEOLS) to what we have set for document.

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/bdb06d5c109f63eac5ab908a7c4f2718fdcc634d/PowerEditor/src/Notepad_plus.rc#L434
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/bdb06d5c109f63eac5ab908a7c4f2718fdcc634d/PowerEditor/src/NppCommands.cpp#L434-L440

I don't know why this command from the `Edit` menu is shown as the same as `Ctrl+V`, since they work differently.
![image](https://user-images.githubusercontent.com/2730894/211403395-4e23b50a-fffc-4923-a4c7-ec640aed405a.png)

Maybe this line should be completely removed:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/bdb06d5c109f63eac5ab908a7c4f2718fdcc634d/PowerEditor/src/NppCommands.cpp#L439

Why inserting even one character should change the EOL in the entire document?

`Paste` in `Edit` menu should have the same behavioru as `Ctrl+V / Shift+INS` and maybe we should have option in Pref (or other dynamically switch) to control [`SCI_SETPASTECONVERTENDINGS`](https://www.scintilla.org/ScintillaDoc.html#SCI_SETPASTECONVERTENDINGS)

Paste with EOL preservation we can always obtain by `Edit > Paste Special > Paste Binary Content` or use `Edit > Clipboard History`

`Paste and convert EOL in whole document` is necessary here as separate command? Personally, I'd prefer to have a command like `Edit > EOL Conversion > Whole document to default`, or just `To default` so it could also work with selection. Thanks to this, if needed, it could be used with other commands (via macros). So to obtain actual behavior we could create macro with `Paste` + `To default`.

Here, anyway, the main problem comes first: should Notepad++ convert EOL itself? If so, should there be any control over it? If we have control and allow mixed EOL then some commands should be tweaked to work properly (lines with different EOL probably should be treated as different).

>@Ekopalypse 
>Nice find, so a config setting like "disallow paste converting eols" and making pasting via the menu consistent with CTRL+V would make almost everyone happy, right !?
>>@donho 
>>Once again, "too many options, no option".

Forcing a conversion isn't always desirable, and it can degrade performance on large files, so this isn't an ideal approach either. In general, automatic modifying the original data without our permission (control over it) will always be a problem at some point.

**Comment by donho** [comment]
> Forcing a conversion isn't always desirable, and it can degrade performance on large files, so this isn't an ideal approach either. In general, automatic modifying the original data without our permission (control over it) will always be a problem at some point.

It makes sense.
The line `_pEditView->execute(SCI_CONVERTEOLS, eolMode);` will be removed.

**Comment by donho** [comment]
Fixed in #14346
